### ISSUE TITLE
Fix bug that prevents GHA "Schedule Monthly" from running 

### ISSUE THREAD

**Issue by t-will-gillis** [hackforla/website#4788]
### Prerequisites
1. Be a member of Hack for LA. If you have not joined yet, please follow the steps on our [Getting Started page](https://www.hackforla.org/getting-started) and attend an onboarding session.
2. Before you claim or start working on an issue, please make sure you have read our [How to Contribute to Hack for LA Guide](https://github.com/hackforla/website/blob/gh-pages/CONTRIBUTING.md).

### Overview
We need our GitHub Actions to run  as expected on the schedule that we have set so that our workflow is not interrupted. The GHA "Schedule Monthly" has failed on the last two scheduled runs and the error needs to be addressed.

### Details
When reviewing HfLA "Actions" logs on the repo, it can be seen that the GHA "Schedule Monthly" has attempted to run on schedule on the 1st of the month for the last three months. Both attempts have failed on the the job `list-inactive-members` on the step `Get List`.
A successful run of this GHA will result in an issue being generated on the HfLA Project Board similar to the following, as well as a notification comment being added to issue #2607. 

- #4567 

### Action Items
Suggested steps:
- [x] Compare the Action logs for the failed GHA on the HfLA repo with the similar logs on the personal repos for hints as to what might be causing the error(s).
- [x] Refactor the [schedule-monthly-PREV.yml](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-monthly-PREV.yml) and js files and/or any other files or settings (such as the tokens) as needed. >NOTE: the file will be renamed `schedule-monthly-PREV.yml` after #4768 is merged.
- [x] Create a demo Project Board and Personal Access Token for your repo- see [HfLAs GitHub Actions](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions) doc esp. Tips 6 and 7. 
  - [x] **Important note: be sure to give your secret token a different name than HfLA's token.**
- [x] Create at least one new issue on your Project Board and remember the number.
- [x] Given the different behavior of the GHA running on a personal repo versus the HfLA repo, add a `workflow_dispatch:` to Line 6 of the schedule-monthly.yml so that the action can be triggered immediately on the live site, i.e.:
```
on:
  schedule:
    - cron: "0 8 1 * *"
  workflow_dispatch:
```
- [x] **_Done_**   ~On_ the yml file, change line 11 `- uses: actions/checkout@v2` to `- uses: actions/checkout@v3` per #4735~
- [x] On the three js files, for testing only change any lines referring to `const owner = "hackforla"` to your own repo. (You will revert this before you open the PR.)
- [x] In [comment-issue.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/trigger-schedule/list-inactive-members/comment-issue.js) change `let agendaAndNotesIssueNumber = 2607;` to reference the issue you created. (You will revert this before you open the PR.)
- [x] Confirm with a test on your own repo. A successful test will create the 'Inactive Member' issue in your repo, and add the notification comment to the issue you designated.

### Resources/Instructions

- Never done GitHub actions? [Start here!](https://docs.github.com/en/actions)
- [Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
- [Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
- [actions/github-script](https://github.com/actions/github-script)
- [schedule-monthly.yml](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-monthly.yml)
- [/list-inactive-members/](https://github.com/hackforla/website/tree/gh-pages/github-actions/trigger-schedule/list-inactive-members)
- [workflow logs](https://github.com/hackforla/website/actions/workflows/schedule-monthly.yml)
- Review issues #4158 and PR #4545 

- Note: `schedule-monthly.yml` is intended to identify and remove "Inactive" members from 'website-write' team, which also addresses ER #4541

**Comment by t-will-gillis** [comment]
Pulling this issue back- it seems that `schedule-thu-1100.yml` (re #4768) and `schedule-monthly.yml` are intending to do the same thing (find members who are inactive) but neither action works currently. We need to decide which action should be kept, and then refactor it so that it runs correctly.

**Comment by t-will-gillis** [comment]
available daily 6-9 pm
eta: 9/8

**Comment by t-will-gillis** [comment]
Progress: the planned features are mostly complete. Having a hang up with accessing an artifact that I want to transfer between GitHub Action jobs. The artifact is successfully uploaded via `actions/upload-artifact@v3`and I can confirm that it has the expected content. The following job is able to download the artifact via `actions/download-artifact@v3` and returns a successful message. But I am thus far unable to access and therefore use the data. 

If I keep the revised GHA as a single job, then we do not need the artifact because the data passes step to step. TBC

**Comment by ExperimentsInHonesty** [comment]
# DRAFT

# Review of Inactive Website Team Members
## Inactive Members
Developers: If your name is on the following list, our team bot has determined that you have been inactive with the Website team in the last 30 days. If we don't hear back from you in the upcoming weeks, we will unassign you from any issues you may be working on and remove you from the 'website-write' team.

cnk
angieyan
ykim25
vanyanv
ericvennemeyer
Jaretzbalba
matthewmpan
Joyce750526
sijiapitts
RyanGehris
DevRishiJain

### Did we make a mistake?
The bot is checking for the following activity:
- If you are assigned to an issue, that you have provided an update on the issue in the past 30 days.  The updates are due weekly.
- If your issue is a draft (in the new issue approval column) that you are adding to it within the last 30 days.
- If you are reviewing PRs, that you have done some kind of activity in the past 30 days.

If you have been inactive in the last 30 days (using the above measurement), you can become active again, by doing one of the above actions.

If you were active during the last 30 days (using the above measurement), the bot did make a mistake, let us know, by responding in a comment (reopening this issue) with this message:
```
I am responding to Issue #[this issue].
The Hack for LA website bot made a mistake, I have been active!  
See my Issue # or PR #  review
```
After you have left the comment, please send us a Slack message to hfla-site channel with a link to your comment.


### Temporary leave
If you have taken a temporary leave, and you have been authorized to keep your assignment to an issue.  
- Your issue is in the "Questions / Review" column, with the `Ready for Dev Lead` label and a note letting us know when you will be back.  We generally try to encourage you to unassign yourself from the issue and allow us to return it to the prioritized backlog.  However, exceptions are sometimes made.

### Removed Members
Our team bot has determined that the following member(s) have not been active with the Website team for over 60 days, and therefore the member(s) have been removed from the 'website-write' team.

cnk

If this is a mistake or if you would like to return to the Hack for LA Website team, please respond in a comment with this message:
```
I am responding to Issue #[this issue].
I want to come back to the team!
Please add me back to the write team, I am ready to work on an issue now.
```
After you have left the comment, please send us a Slack message to hfla-site channel with a link to your comment.

**Comment by ExperimentsInHonesty** [comment]
@t-will-gillis I left a revised language that should be used in the issue (in the comment above).  You will notice that
- I didn't put the @ symbol in front of their names.  You can't do that until we go live, because otherwise it messages them.  
- I have in the draft a "[this issue]" text.  I don't know if it's possible, but if it is, possible to reference the issue number in the draft comment we are giving them, that would be good.

Notes 
- GHA should check all columns except for QA, Done, Questions / Review.
- When checking PRs that it checks all reviewers activity and submitters activity
- When checking issues with the DRAFT label, that it checks the Top comment (body) for changes.
- Role label should be role: dev leads
- Feature labels should be: `feature: Administrative` and `Feature: Onboarding/Contributing`
- Can we add a milestone to the issue automatically (Team workflow)?
- Can we add it to the project board automatically?
- Can we close this issue as soon as it's created?
- Please specify work flow when the user writes back to us.  What we have to do at each juncture
   - When the bot makes a mistake on 30 inactive
   - When someone wants to come back after the 60 day removal

**Comment by t-will-gillis** [comment]
@ExperimentsInHonesty  Please cross reference my repo's [issue 359](https://github.com/t-will-gillis/website/issues/359)
> @t-will-gillis I left a revised language that should be used in the issue (in the comment above). You will notice that
>
>- I didn't put the @ symbol in front of their names. You can't do that until we go live, because otherwise it messages them.
>- I have in the draft a "[this issue]" text. I don't know if it's possible, but if it is, possible to reference the issue number in the draft comment we are giving them, that would be good.  
- WG: _I am leaving a space/ gap between the "@" and the person's name so that the person is not messaged until this goes live._
- WG: _Regarding the "this issue" number: currently, the data for the body of the issue (which includes the line that references "this issue") is being sent to "github.rest.issues.create()" first, and then the issue number is being assigned afterwards._ ~_So as written currently, we cant reference the issue number. But there may be one or two things I can try, such as commenting after the issue is created etc._~ UPDATE: _I added a query to return all active issues in the repo, limiting the result to one which returns the last issue created. Then I am adding 1 to get the number of the next issue, i.e. the one about to be generated. Seems to be working!_

Notes

- GHA should check all columns except for QA, Done, Questions / Review.
  - WG: _the GHA currently checks for **all** activity, any column. I think this is a good thing b/c otherwise the GHA might ignore recent contributions to an issue that was very recently closed????_
- When checking PRs that it checks all reviewers activity and submitters activity
  - WG: _I ran a test for the previous 12 hours. I believe the GHA is picking up all reviewers and all submitters._
- When checking issues with the DRAFT label, that it checks the Top comment (body) for changes.
  - WG: _I made test comments to this issue to confirm: Currently the action does log that I edited the top comment, as well as edited this comment and added a new comment._
- ~Role label should be role: dev leads~
  - WG: _**Done**_ 
- ~Feature labels should be: feature: Administrative and Feature: Onboarding/Contributing~
  - WG: _**Done**_ 
- ~Can we add a milestone to the issue automatically (Team workflow)?~
  - WG: ~I have tried to add the milestone previously, but I haven't found a way to do this (yet).~
    UPDATE:  Woohoo! _Yes, after a painful search I figured out the problem: the API wants an integer and not a string, and the correct integer is based on the order of when each workflow was created. The correct integer is 8- b/c coincidentally '.08 Team workflow' was the 8th milestone created._
- ~Can we add it to the project board automatically?~
  -  WG:_**Yes**- it automatically adds issue to the project board_   
- ~Can we close this issue as soon as it's created?~
  - WG:_I have tried but haven't figured out a way to close the issue while it is being opened this yet._ 
- Please specify work flow when the user writes back to us. What we have to do at each juncture
  - When the bot makes a mistake on 30 inactive
  - When someone wants to come back after the 60 day removal
    - WG: _If the bot seems to be making more than one or two mistakes, I think at that point if we think it is a problem we would want to take it down to figure out what is going wrong. Otherwise, for the person who wants to come back after 60 days, wouldn't an Admin add them back? Does the person need to be re-onboarded?_
    
 WG: For an issue written many months ago that is being worked on now, it appears that the GHA might be including the original author in the group of "active contributors"....

**Comment by t-will-gillis** [comment]
TESTING

**Comment by roslynwythe** [comment]
@t-will-gillis @ExperimentsInHonesty Regarding instructions for the members flagged as "inactive 30 days", I wonder if we should have them leave a comment in their prework issue instead of this issue, so that the comment itself will count as activity.

**Comment by roslynwythe** [comment]
Hi @t-will-gillis @ExperimentsInHonesty I just wanted to confirm that if the user flagged as "30 days inactive" follows the instructions and creates a comment in the "review inactive team members" issue, it will count as activity, so they will not be flagged as inactive in the next montly run.  So if that is the case, then the dev lead doesn't have to do anything to resolve the situation, except figure out why the dev was incorrectly deemed as inactive.  

Whereas if a developer is incorrectly removed after 60 days of inactivity, the dev lead has to figure out why the dev was incorrectly deemed as inactive and also has to restore membership in website-write.  For a member who has lost membership in website-write, and only has "read" access to the repository, I don't think they could reopen an closed issue that was created by another user [based on repo permission chart](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization), so that would mean we would have to keep the "review inactive team members" issue open.  Perhaps it would be better to have them write the comment in their prework checklist instead.  Or just skip the instruction to add the comment, and have them leave a Slack message.  Also I wonder if we should have them leave the message in #hfla-site-merge instead of #hfla-site since only the merge team can help them.

**Comment by t-will-gillis** [comment]
> I just wanted to confirm that if the user flagged as "30 days inactive" follows the instructions and creates a comment in the "review inactive team members" issue, it will count as activity, so they will not be flagged as inactive in the next montly run. So if that is the case, then the dev lead doesn't have to do anything to resolve the situation, except figure out why the dev was incorrectly deemed as inactive.

_WG: Correct- if the "30 day inactive" member posts a comment, review, opens an issue, etc. this will count as activity and the bot should skip them the following month._
_WG: If there are a lot of mistakes/ many people incorrectly being placed on the inactive list, then someone (= me) configured something incorrectly and I think the best course of action is to deactivate the GHA until it can be fixed._


> Whereas if a developer is incorrectly removed after 60 days of inactivity, the dev lead has to figure out why the dev was incorrectly deemed as inactive and also has to restore membership in website-write. 

_WG: Agree_

> For a member who has lost membership in website-write, and only has "read" access to the repository, I don't think they could reopen an closed issue that was created by another user [based on repo permission chart](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization), so that would mean we would have to keep the "review inactive team members" issue open. Perhaps it would be better to have them write the comment in their prework checklist instead. 

_WG: Yes- you're correct they would not be able to comment, and I agree we should remove the reference to adding the comment b/c they won't be able to. They probably would not be able to comment on their prework either._

>Or just skip the instruction to add the comment, and have them leave a Slack message. Also I wonder if we should have them leave the message in #hfla-site-merge instead of #hfla-site since only the merge team can help them.

_WG: Agree with both statements_

**Comment by ExperimentsInHonesty** [comment]
1. Can we close this issue as soon as it's created?
     - WG:I have tried but haven't figured out a way to close the issue while it is being opened this yet.
     -  BW: second cron that runs after.
1. Please specify work flow when the user writes back to us. What we have to do at each juncture
   - We decided (Roslyn, Will, Jessica, Bonnie) that if the text is any of the following it should reopen the issue and puts it in the questions and review column with a `ready for dev lead` label.
      - "The Hack for LA website bot made a mistake"
      - "I want to come back to the team"
1. WG: For an issue written many months ago that is being worked on now, it appears that the GHA might be including the original author in the group of "active contributors"....
   - authors of issues should be counted as active if the issue was created within the time window.

**Comment by t-will-gillis** [comment]
Hello @ExperimentsInHonesty and @roslynwythe The following are done. We probably should discuss item 2...

Updates:
1. Created [wr-schedule-monthly.yml](https://github.com/t-will-gillis/website/blob/fix-sch-monthly-TEST-10-22/.github/workflows/wr-schedule-monthly.yml) workflow that is triggered by a successful 'Schedule Monthly' run.
  
2. Regarding exploring ways to reopen the issue upon an issue comment, then add label and move it. 
    - I have an alternate [schedule-monthly-reopen.yaml](https://github.com/t-will-gillis/website/blob/fix-sch-monthly-TEST-10-22/.github/workflows/schedule-monthly-reopen.yaml) that depends on the dev commenting and then reopening the issue. Once they reopen the issue, then the issue moves to "Questions / In Review" and the `ready for dev lead` label is applied. I believe that having the dev select to "Reopen with comment" is preferable since this workflow will only be triggered when an issue is reopened.
    - The workflow that automatically reopens the issue is here: [schedule-monthly-reopen.yml](https://github.com/t-will-gillis/website/blob/fix-sch-monthly-TEST-10-22/.github/workflows/schedule-monthly-reopen.yml) using "on: issue_comment":
      - reopens the issue if a comment is added that contains 'bot made a mistake' or 'want to come back',
      - adds the `ready for dev lead` label
      - moves it into the "Questions / In Review " column. 
    - Again, I think the first workflow is preferable because although the latter GHA only runs under the specific condition, it will still be triggered whenever anyone makes any comment on any issue.
 4. Regarding whether activity extends to the authors of issues: 
    - _**UPDATE 11/22/23:**_ _I have reworked the API call in_ `contributors-data.js` _that should only select issue_ `assignee`s.  
    
    - ~I tested `contributors-data.js` checking activity over the past 27 hours- here are two issue-authors with activity:~
      - ~Micah authored issue 5825. Issue 5876 is listed as a dependent, and itself was updated yesterday which registered as an update to the issue `body`, and credit for activity is given to Micah.  I don't see an easy way to separate the activity- the activity itself is not credited to an 'author'. I.e. there does not seem to be a way to distinguish between Micah making an edit, or the bot or someone else making an edit (that I can see).~
      - ~Similar with Justin.~
    - ~This discrepancy is likely due to the choice of api calls on the original `contributors-data.js`- i.e. these need to be refined so that the original author is not included when someone else updates their issue.~

 5. Added (to the [Inactive Team Members Issue](https://github.com/t-will-gillis/website/issues/363)) a short description of what dev lead needs to do in case someone needs to be added back.

**Comment by t-will-gillis** [comment]
@roslynwythe @ExperimentsInHonesty 

(from previous comment above)
2. continuing - In the 11/27 Monday meeting, we discussed the scenario where a member who did not have write access would not have permission to use the "Reopen with comment" option, thus this option would not work. We also discussed that the second option for triggering the original issue to reopen using "on: issue_comment" will mean that this GHA will start then stop for every comment made on every issue, and so this also would not work.

The alternate option that I believe will work is to have the member comment on the "Review Inactive Team Members" issue, and then open a new issue from the comment itself. This option works for any member regardless of write status. I was able to confirm this in practice by using the same method to open an issue through a 'dummy' account: #[5939](https://github.com/hackforla/website/issues/5939) 

When the member opens a new issue this way, the issue passes through existing GHAs, and with minor edits the GHAs can apply appropriate labels and move the issue to the "Questions  / In Review" column.

The details with examples are described in #5467 under the section "11/29/23: Final Addendum".
